File: java/iroha_android/src/main/java/org/hyperledger/iroha/android/connect/error/ConnectErrorTelemetryOptions.java
package org.hyperledger.iroha.android.connect.error;

/** Optional overrides when rendering telemetry attributes. */
public record ConnectErrorTelemetryOptions(
    Boolean fatal, Integer httpStatus, String underlying) {

  public ConnectErrorTelemetryOptions {
    // Defensive copies are unnecessary; fields are immutable.
  }

  public static ConnectErrorTelemetryOptions empty() {
    return new ConnectErrorTelemetryOptions(null, null, null);
  }

  public ConnectErrorTelemetryOptions withFatal(Boolean value) {
    if ((fatal == null && value == null)
        || (fatal != null && fatal.equals(value))) {
      return this;
    }
    return new ConnectErrorTelemetryOptions(value, httpStatus, underlying);
  }
}

File: kotlin/core-jvm/src/main/java/org/hyperledger/iroha/sdk/connect/error/ConnectErrorTelemetryOptions.kt
package org.hyperledger.iroha.sdk.connect.error

/** Optional overrides when rendering telemetry attributes. */
data class ConnectErrorTelemetryOptions(
    @JvmField val fatal: Boolean?,
    @JvmField val httpStatus: Int?,
    @JvmField val underlying: String?,
) {
    companion object {
        @JvmStatic
        fun empty(): ConnectErrorTelemetryOptions =
            ConnectErrorTelemetryOptions(null, null, null)
    }

    fun withFatal(value: Boolean?): ConnectErrorTelemetryOptions {
        if ((fatal == null && value == null) || (fatal != null && fatal == value)) {
            return this
        }
        return ConnectErrorTelemetryOptions(value, httpStatus, underlying)
    }
}
